drape_frontend/stylist.cpp

```diff
@@ -265,90 +265,49 @@ bool IsHatchingTerritoryChecker::IsMatched(uint32_t type) const
+                              feature::GeomType const type, bool const auxCaptionExists)
 {
   feature::NameParamsOut out;
-  if (auxCaptionExists || type == feature::GeomType::Line)
+  // TODO: remove forced secondary text for all lines and set it via styles for major roads and rivers only.
+  // ATM even minor paths/streams/etc use secondary which makes their pathtexts take much more space.
+  if (zoomLevel > scales::GetUpperWorldScale() && (auxCaptionExists || type == feature::GeomType::Line))
+  {
+    // Get both primary and secondary/aux names.
     f.GetPreferredNames(true /* allowTranslit */, deviceLang, out);
+    m_auxText = out.secondary;
+  }
   else
-    f.GetReadableName(true /* allowTranslit */, deviceLang, out);
-
-  m_mainText = out.GetPrimary();
-  m_auxText = out.secondary;
-
-  // Set max text size to avoid VB/IB overflow in rendering.
-  size_t constexpr kMaxTextSize = 200;
-  if (m_mainText.size() > kMaxTextSize)
-    m_mainText = m_mainText.substr(0, kMaxTextSize) + "...";
-
-  m_houseNumber = f.GetHouseNumber();
-
-  ProcessZoomLevel(zoomLevel);
-  ProcessMainTextType(mainTextType);
-}
-
-std::string const & CaptionDescription::GetMainText() const
-{
-  return m_mainText;
-}
-
-std::string const & CaptionDescription::GetAuxText() const
-{
-  return m_auxText;
-}
-
-bool CaptionDescription::IsNameExists() const
-{
-  return !m_mainText.empty() || !m_houseNumber.empty();
-}
-
-void CaptionDescription::ProcessZoomLevel(int const zoomLevel)
-{
-  if (zoomLevel <= scales::GetUpperWorldScale() && !m_auxText.empty())
   {
-    m_auxText.clear();
+    // Returns primary name only.
+    f.GetReadableName(true /* allowTranslit */, deviceLang, out);
   }
+  m_mainText = out.GetPrimary();
+  ASSERT(!m_auxText.empty() && !m_mainText.empty() || m_auxText.empty(), ("auxText without mainText"));
 
-  if (zoomLevel < 5 && m_mainText.size() > 50)
+  uint8_t constexpr kLongCaptionsMaxZoom = 4;
+  size_t constexpr kLowWorldMaxTextSize = 50;
+  if (zoomLevel <= kLongCaptionsMaxZoom && m_mainText.size() > kLowWorldMaxTextSize)
   {
     m_mainText.clear();
     m_auxText.clear();
+    return;
   }
-}
 
-void CaptionDescription::ProcessMainTextType(drule::text_type_t const & mainTextType)
-{
-  if (m_houseNumber.empty())
-    return;
+  // Set max text size to avoid VB/IB overflow in rendering.
+  size_t constexpr kMaxTextSize = 200;
+  if (m_mainText.size() > kMaxTextSize)
+    m_mainText = m_mainText.substr(0, kMaxTextSize) + "...";
 
-  if (mainTextType == drule::text_type_housenumber)
-  {
-    /// @todo this code path is never used, probably need to have e.g. "text: housenumber" in styles for it.
-    m_mainText.swap(m_houseNumber);
-    m_houseNumber.clear();
-    m_isHouseNumberInMainText = true;
-  }
-  else if (mainTextType == drule::text_type_name)
+  uint8_t constexpr kHousenumbersMinZoom = 16; // hardcoded for optimization
+  if (zoomLevel >= kHousenumbersMinZoom && (auxCaptionExists || m_mainText.empty()))
```

Comment:
Looks like a good optimization. Better not to hardcode, but take out for example into RulesHolder and calculate this scale once. Like RulesHolder::GetMinHNZoom().

Reply:
Yes, it'll be more flexible for sure.
But it'll also pollute RulesHolder loading logic with the housenumbers exception. So instead of a single line hardcoded housenumbers zoom level we'll have many hardcoded lines for a way to get this housenumbers zoom level...
And we have a hardcoded z18- min zoom for point housenumbers anyway....
So I just left a TODO for now.